Jetpack plugin: Show discount in Assistant

monsieur-z opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
This PR shows the timed discount in the plugin Assistant, for eligible sites.

Information about prices and discounts is fetched from the API. This PR builds the UI.

A couple of notes:
- A couple of new UI elements were added: `BackButton`, and `DiscountBadge`
- A new high order component was created to easily provide the URL to checkout for upsold products: `withUpgradeUrl`
- The suggestions screen is no longer a step of the Assistant wizard

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
Context: pbNhbs-1xX-p2 

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
This PR adds a couple of new tracking events.

When displaying/clicking the upsell card (with a `discount: boolean` prop):
```
jetpack_recommendations_upsell_card_display
jetpack_recommendations_upsell_card_cta_click
```

When displaying/clicking a product card from the product suggestion screen (with `discount: boolean` and `product_slug: string` props):
```
jetpack_recommendations_product_suggestion_display
jetpack_recommendations_product_suggestion_click
```

When displaying/clicking a product card from the summary screen (with `discount: boolean` and `product_slug: string` props):
```
jetpack_recommendations_summary_sidebar_display
jetpack_recommendations_summary_sidebar_click
```

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

**Prerequisites**
- Download this branch and launch your testing site, or spin up a new JN site with this branch.
- Apply D79054-code to your sandbox, and follow its instructions.
- Make sure your testing site isn't connected to Jetpack and has no subscriptions.

_Note: to reset the Assistant, run the following:_
```
jetpack docker wp jetpack options delete recommendations_data
jetpack docker wp jetpack options delete recommendations_step
```

**Instructions**
- Connect your site to Jetpack, and choose the Free plan.
- At this point, a coupon must have been created for your site.
- Visit the Assistant at `/wp-admin/admin.php?page=jetpack#/recommendations`.
- You should see the follow screen. Make sure the countdown makes sense (offer expires after 7 days).
<img width="1013" alt="Screen Shot 2022-04-19 at 6 37 13 PM" src="https://user-images.githubusercontent.com/1620183/164112672-e73a35fe-f7fe-4006-b8d9-a98ef7bd46d6.png">

- Then, click on _View discounted products_. Make sure you see the following screen:
<img width="1009" alt="Screen Shot 2022-04-19 at 8 11 31 PM" src="https://user-images.githubusercontent.com/1620183/164121190-edf94825-aad6-42ca-ac88-60ef34652ac1.png">

- Check that:
  - For Backup and Security, clicking the CTA leads you to the checkout page with the discount applied.
  - Prices and discounts in the Assistant and the checkout page match.
  - The highest discount percentage of the 2 products matches the discount percentage in the previous screen.
  - The countdown still makes sense.
- Return to the Assistant first screen by hitting the back button.
- Complete or skip the Assistant steps. Check that you see the Security product card in the summary:
<img width="1008" alt="Screen Shot 2022-04-19 at 8 38 05 PM" src="https://user-images.githubusercontent.com/1620183/164123256-b79d7099-2e24-43b6-8c10-61b1ec1a3233.png">

- Verify the links, discounts, prices, and timer like before.


**Alternate scenarios**
- **Discount disabled**: disable site discount in D79054-code. You should still see the upsell and product cards, but without discount.
<img width="1006" alt="Screen Shot 2022-04-20 at 3 32 59 PM" src="https://user-images.githubusercontent.com/1620183/164564248-f8443cb9-33b4-4a6a-8eeb-4500d8c26d79.png">
<img width="1002" alt="Screen Shot 2022-04-20 at 3 39 52 PM" src="https://user-images.githubusercontent.com/1620183/164564281-a79d1197-d94f-45a6-987d-269092751ecb.png">

- **Site with any subscription**: you should see neither upsell nor discount.
<img width="1008" alt="Screen Shot 2022-04-20 at 3 17 46 PM" src="https://user-images.githubusercontent.com/1620183/164564398-48ee5faf-b158-4faf-8837-bfdb00fe9089.png">

- **Site with no discount**: delete the coupon code in the site blog options. You should see no discount.
- **Site with expired coupon**: update the expiry date in the dedicated tool. You should see no discount.
- **Site with used coupon**: try reusing a coupon. You should see no discount.
- **From dashboard banner**: from the Assistant banner in the WPadmin dashboard, click _Continue_ and make sure you then see the upsell card with the discount.
<img width="1103" alt="Screen Shot 2022-04-21 at 7 03 03 PM" src="https://user-images.githubusercontent.com/1620183/164565004-cefcd909-34af-4520-a593-2b4a51556c51.png">

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/index.jsx:
```diff
@@ -96,6 +98,8 @@ const RecommendationsComponent = props => {
 			<QueryRewindStatus />
 			<QuerySite />
 			<QuerySitePlugins />
+			<QuerySiteDiscount />
```
Both sets of data allows us to compute the final price and total discount.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/product-card-upsell/index.jsx:
```diff
@@ -1,56 +1,81 @@
+	introOffers,
+	upgradeUrl,
 } ) => {
-	useEffect( () => {
```
Moved to ancestor and passed as `onMount` prop.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/product-card-upsell/index.jsx:
```diff
@@ -1,56 +1,81 @@
-		} );
-	}, [ product_slug ] );
-
-	const onUpsellClick = useCallback( () => {
```
Moved to ancestor and passed as `onClick` prop.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/product-suggestions/index.jsx:
```diff
@@ -1,134 +1,94 @@
-		suggestions,
-	} = props;
-
-	useEffect( () => {
```
Though still in the `prompts` folder for now, this section is not a step of the wizard anymore. This explains the removed code.

monsieur-z replied:
Yes, it's another way to look at it. I guess we can leave it in that folder then.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/prompt-layout/index.jsx:
```diff
@@ -51,37 +92,8 @@ const PromptLayout = props => {
 				{ content }
 				<div className="jp-recommendations-question__answer">{ answer }</div>
 			</div>
-			{ illustrationPath && ! sidebarCard && (
```
Moved to `SideContent`.

jboland88 commented on projects/plugins/jetpack/_inc/client/recommendations/sidebar/discount-card/index.jsx:
```diff
@@ -54,26 +71,39 @@ const DiscountCard = ( { discount, expiryDate } ) => {
-						expiryDate={ expiryDate }
-					/>
-				</div>
+				{ hasDiscount && (
```
Should we also check for `expiryDate` here?

monsieur-z replied:
`hasDiscount` will be false is `expiryDate` is invalid or passed.

jboland88 commented on projects/plugins/jetpack/_inc/client/recommendations/product-card-upsell/index.jsx:
```diff
@@ -65,48 +90,70 @@ const ProductCardUpsell = ( {
-					{ `${ cost_timeframe }, ${ billing_timeframe }` }
-				</div>
-			</div>
+			{ ! ( isFetchingDiscount || isFetchingOffers ) && (
```
If this data is still loading, is there a loading state that can be shown?

monsieur-z replied:
I've added handling this loading state in a follow-up task.

jboland88 reviewed (changes requested):
I have not had a chance to do functional testing yet, but have added some comments.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/utils/compute-max-suggested-discount.js:
```diff
@@ -0,0 +1,40 @@
+
+			return Math.round( totalDiscount * 100 );
+		} )
+		.filter( Boolean );
```
That's it. If the array had an undefined value, `Math.max` would return `NaN`.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/product-card-upsell/index.jsx:
```diff
@@ -1,56 +1,81 @@
+			onMount();
+		}
+	}, [] );
+
```
> Something about having an onMount prop feels odd
Just curious, do you know why it feels like this to you?

I've gone through several rounds of refactoring there. Taking a look back at it now, it seems that the events can be tracked directly in the cards ancestors indeed 👍

jboland88 replied:
I think it feels like that because it seems like if there is some logic that should be executed when this particular component is mounted, it should live in the component itself instead of being passed in from somewhere else. IMO it makes the code a bit harder to follow since the reader needs to be aware of the context where this component is used to know what happens when it mounts. 

I don't think that's a rule though, just my preference.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/product-suggestions/index.jsx:
```diff
@@ -1,134 +1,92 @@
+	discountData,
+} ) => {
+	const { discount, expiry_date: expiryDate, is_used: isUsed } = discountData;
+	const hasDiscount = useMemo(
+		() => discount && ! isUsed && new Date( expiryDate ).valueOf() - Date.now() > 0,
+		[ discount, isUsed, expiryDate ]
```
Even more though since I've added tracking events. I've moved the logic to a function.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/product-suggestions/index.jsx:
```diff
@@ -1,134 +1,92 @@
 
-	// Redirect the user to the next step if they are not eligible for the product
-	// suggestions step. We need to check this for the individual step because:
-	// 1. A user can access the step directly through the URL with the current
-	//    implementation of the Recommendations routes.
```
I don't think it was much of a concern indeed, but I've added back the redirection nonethless.

monsieur-z commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/site-type/index.jsx:
```diff
@@ -106,6 +109,7 @@ const SiteTypeQuestionComponent = props => {
 				'jetpack'
 			) }
 			answer={ answerSection }
+			isLoadingSideContent={ isFetchingSuggestions }
```
Background is decided in `PromptLayout`, I believe we still need to pass the prop here if that was your point.

jboland88 replied:
I was thinking that since `<DiscountCard />` already has a check for `isLoading`:

```
const isLoading = isFetchingDiscount || isFetchingSuggestions || isFetchingOffers;
```

That the `isLoading` value in `<DiscountCard />` could be used within the `render` method of `<DiscountCard />` to show some alternate output if the data needed for the card is not yet available. Then, there would not be as much need for this `isLoadingSideContent` prop, which does similar data availability checks - just higher up in the component tree.

monsieur-z replied:
The card container does need that information too, though. But maybe less so one the illustrations have been updated. Let's revisit that later.

monsieur-z commented:
> Side note: Maybe it was a caching thing or something, but FWIW, for a while I kept getting taken to `/wp-admin/admin.php?page=jetpack#/recommendations/videopress` when trying to visit `/wp-admin/admin.php?page=jetpack#/recommendations`. After a few hard refreshes, it started working.

@travisw: yeah, I suspect there's some kind of caching issue too. When I reset options during dev, it usually takes a little while before it becomes effective.

monsieur-z commented:
> The checkout defaults to showing the yearly price, which doesn't match what's shown in the assistant (_probably_ OK).
> Switching to monthly pricing now shows the undiscounted amount from the assistant, discounted with the new coupon. Like we're applying this new 10% discount but not the original (60%?) first year price discount. This results in the checkout showing a higher price than the assistant. It's possible I've misunderstood something with the project.

Thanks for spotting that, @travisw. When switching from the yearly to the monthly term in the checkout page, the 10% discount should not apply anymore indeed. I'd have expected the coupon to get removed, since it is only valid for yearly Backup and Security.

I'll check how we can solve this, but I doubt the fix will be in this codebase.

jboland88 commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/product-suggestions/index.jsx:
```diff
@@ -1,134 +1,108 @@
-					{ __( 'Decide later', 'jetpack' ) }
-				</a>
-			</div>
-			<div className="jp-recommendations-product-suggestion__footer">
-				{ __( 'Special introductory pricing, all renewals are at full price.', 'jetpack' ) }
-			</div>
```
I think this text needs to be retained when `hasDiscount` is false.

jboland88 commented on projects/plugins/jetpack/_inc/client/recommendations/summary/upsell.jsx:
```diff
@@ -0,0 +1,89 @@
+			) }
+			<div className="jp-recommendations-summary__footer">
+				<MoneyBackGuarantee text={ __( '14-day money-back guarantee', 'jetpack' ) } />
+				{ hasDiscount && (
+					<div className="jp-recommendations-summary__footnote">
+						{ __(
+							'* Discount is for first term only, all renewals are at full price.',
+							'jetpack'
+						) }
+					</div>
+				) }
```
I think we need to show this whether or not there is a discount present, a similar message shows now without the discount:

"Special introductory pricing, all renewals are at full price."

jboland88 reviewed (changes requested):
I have not completed the functional testing just yet.
Had a few more comments - looking good overall.

jboland88 commented:
Main test scenario works well. I noted in some of the inline comments, but I think we need to keep the "Special introductory pricing, all renewals are at full price." text on the product suggestions screen and the summary screen.

I tested completing a purchase with the coupon, which was successful. However, I was just redirected back to `admin.php` on my site, which just showed a white screen.

**Alternate scenarios**

- Discount disabled: This worked, I did not see the discount/ timer ✅  
- Site with any subscription: This did not work completely - I added a backup subscription, and I no longer see the product cards on the `site-type` or `summary` steps, but the product suggestions step is still showing the discounted rates and discount percentage (also, I can access the product suggestions step directly via the URL) ❌ 
- Site with no discount: This seemed to work, I saw the `<DiscountCard />` and product cards on the product suggestions and summary screens with no discount percentage, timer or discounted rate. I did notice, however, that the call to `jetpack/v4/site/discount` is still returning some data - is that correct?
- Site with expired coupon: This did not work as expected, I still see the discounted rate and discount badge showing on the summary screen and product suggestions screen. ❌ 
- Site with used coupon: This did not work as expected, I do not see the discount on the site type screen, but am still seeing the discounted rate and discounted prices on the product suggestion and summary screens. ❌ 
- From dashboard banner: from the Assistant banner in the WPadmin dashboard, click Continue and make sure you then see the upsell card with the discount. - This worked for the most part. I had a difficult time getting the assistant banner to show up again, so I cleared the `recommendations_data` option and then navigated directly to `wp-admin/admin.php?page=jetpack#/recommendations/monitor`. The Discount card did show up there. However, this works for any step and would potentially show the discount card on any recommendations step if the page is accessed directly (with a fresh react state). I'm not sure that this is the original intent? ❓

monsieur-z commented:
Thanks for the thorough review @jboland88!

> Main test scenario works well. I noted in some of the inline comments, but I think we need to keep the "Special introductory pricing, all renewals are at full price." text on the product suggestions screen and the summary screen.

Yes, that makes sense.

> I tested completing a purchase with the coupon, which was successful. However, I was just redirected back to admin.php on my site, which just showed a white screen.

My bad, I introduced an error in the redirect URL. It should be fixed now.

> Site with any subscription: This did not work completely - I added a backup subscription, and I no longer see the product cards on the site-type or summary steps, but the product suggestions step is still showing the discounted rates and discount percentage (also, I can access the product suggestions step directly via the URL) ❌

I can't reproduce the issue here. When I try to access the suggestions page by typing the URL directly in the browser, I'm redirected to the wizard as expected for a site with Backup. Could you provide more details about your testing steps?

> Site with no discount: This seemed to work, I saw the <DiscountCard /> and product cards on the product suggestions and summary screens with no discount percentage, timer or discounted rate. I did notice, however, that the call to jetpack/v4/site/discount is still returning some data - is that correct?

It's normal that the endpoint is called and returns a 200 status code. What payload do you see?

> Site with expired coupon: This did not work as expected, I still see the discounted rate and discount badge showing on the summary screen and product suggestions screen. ❌

I can't reproduce the issue either. I updated the expiry date in the dedicated tool. Is that what you did as well? Or did you let the timer expire in the front end (in which case it's okay to still see the discount—coupon won't work at checkout)?

> Site with used coupon: This did not work as expected, I do not see the discount on the site type screen, but am still seeing the discounted rate and discounted prices on the product suggestion and summary screens. ❌

Good catch! Should be fixed now.

> However, this works for any step and would potentially show the discount card on any recommendations step if the page is accessed directly (with a fresh react state). I'm not sure that this is the original intent? ❓

Yes, it is.

monsieur-z commented on projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js:
```diff
@@ -246,7 +264,7 @@ const stepToNextStep = {
 	'banner-completed': 'woocommerce',
 	'not-started': 'site-type-question',
-	'site-type-question': 'product-suggestions',
+	'site-type-question': 'woocommerce',
```
Yes, I wanted to make sure I didn't break any flow and duplicated the value of `product-suggestions`. It falls back to `monitor`.
